README.rst
=================================
Fetchable: Python Client Side SDK
=================================

This is the Fetchable client-side Software Development Kit (SDK) for Python. It allows developers of robots, smart speakers, voice assistants, etc. to quickly develop programs that make calls to the Fetchable API. The library has only been tested on Ubuntu systems so far.


Table of Contents
=================

-  `Installation <#installation>`__
-  `Usage <#usage>`__
-  `Errors <#errors>`__
-  `Contributing <#contributing>`__
-  `License <#license>`__

Installation
============

Pre-requisites
--------------

* Python 2.7 or 3.6
* An active Fetchable account.

Install Package
---------------

The recommended way to install this package is through pip (or pip3) as follows:

.. code-block:: sh

  $ pip install fetchable


Set Up Keys
-----------

1. Log into your account and on the console page, download the authentication keys and note the location it downloaded to. The file should be named 'fetchable_auth_keys.json'. Note: be sure to keep this file secure and share it or commit it to any source control.

2. By default, when the library runs it will attempt to read the file containing your keys from an environment variable which
specifies the path to the file (you can also override the environment variable by explicitly setting the path in the constructor). Set the environment variable as so:

.. code-block:: sh

  $ export FETCHABLE_AUTH_FILE=/path/to/fetchable_auth_keys.json


Usage
=====
Once you have completed the installation, you can start fetching data from the index. The main class used to interact with the API is the FetchableClient object.

Constructing the Client Object
------------------------------

There are three optional parameters which can be passed to the constructor of the FetchableClient object.

1. API version

The verison of the API to be used can be specified with the :code:`api_version` parameter. The parameter is an enum defined in the configuration file which needs to be imported. The FetchableClient object will default to the latest version in the absence of this parameter.

.. code-block:: python

  from fetchable import FetchableClient
  from fetchable import configuration

  client = FetchableClient(api_version=configuration.api_version.v0_1)

2. User agent

The FetchableClient object allows you to set your own user-agent header, this can be done through the :code:`user_agent` parameter in the format :code:`<custom-user-agent>/<version>` where :code:`version` takes the form  :code:`0.1.0`, :code:`0.2.0`, :code:`1.0.0`, etc. By setting this to something relevant to only your application (e.g. :code:`my-amazing-chatbot/0.1`) and by whitelisting only that value through the console you can enhance the security of your account and distinguish between different applications. The default is value is :code:`fetchable-python-client/x.y.z`.

.. code-block:: python

  from fetchable import FetchableClient

  client = FetchableClient(user_agent='my-amazing-chatbot/0.1.0')


3. Authentication credentials

By default, the libary will attempt to read your authentication keys from the file specified by the :code:`FETCHABLE_AUTH_FILE` environment variable. You can also specify this through the constructor. If both are set, the value passed through to the constructor will override the  environment variable. If neither are set, an exception is thrown.

.. code-block:: python

  from fetchable import FetchableClient

  client = FetchableClient(auth_file='/path/to/file/here.json')



Fetching from endpoints
-----------------------

Once the installation has been completed and the object has been constructed, you can begin making calls against the API endpoints. Note: these endpoints only cover version v0.1 of the API, visit the `endpoint <https://fetchable.ai/docs/api/endpoints>`_ documentation for more details on these.

1. API status endpoint

This endpoint is useful to test the connection and authentication of the client as well as receive the current status of the API.

.. code-block:: python

  from fetchable import FetchableClient

  client = FetchableClient()

  status_response = client.status()

  if(status_response['status_code']==200):
      print("The Fetchable API is up - all systems are go")
  elif(status_response['status_code']==1001):
      print("I can't connect to the internet right now...")
  else:
      print(status_response)

2. Entity-attribute endpoint

This endpoint is used to fetch the attributes of entities in our index.

.. code-block:: python

  from fetchable import FetchableClient

  client = FetchableClient()

  entity_response = client.fetch_entity_attribute("Mount Everest", "Elevation")

  if(entity_response['status_code']==200):
      print("The height of mount_everest is {} {}.".format(entity_response['value'], entity_response['unit']))
  elif(entity_response['status_code']==1001):
      print("Can't connect to the internet right now...")
  else:
      print(entity_response)

3. Dictionary endpoint

Used to fetch the definitions of words.

.. code-block:: python

  from fetchable import FetchableClient

  client = FetchableClient()

  definition_response = client.fetch_word_definition("ameliorate")

  if(definition_response['status_code']==200):
      print("The definition of {}: is {}".format("ameliorate", definition_response['meanings'][0]))
  elif(definition_response['status_code']==1001):
      print("Can't connect to the internet right now...")
  else:
      print(definition_response)


4. Joke endpoint

.. code-block:: python

  from fetchable import FetchableClient

  client = FetchableClient()

  joke_response = client.fetch_joke()

  if(joke_response['status_code']==200):
      print("{} - {}".format(joke_response['setup'], joke_response['punchline']))
  elif(joke_response['status_code']==1001):
      print("Can't connect to the internet right now...")
  else:
      print(joke_response)

5. Inspirational quote endpoint

.. code-block:: python

  from fetchable import FetchableClient

  client = FetchableClient()

  quote_response = client.fetch_quote()

  if(quote_response['status_code']==200):
      print("{} by {}".format(quote_response['quote'], quote_response['author']))
  elif(quote_response['status_code']==1001):
      print("Can't connect to the internet right now...")
  else:
      print(quote_response)

4. Fun fact endpoint

.. code-block:: python

  from fetchable import FetchableClient

  client = FetchableClient()

  fun_fact_response = client.fetch_fun_fact()

  if(fun_fact_response['status_code']==200):
      print(fun_fact_response['fun_fact'])
  elif(fun_fact_response['status_code']==1001):
      print("Can't connect to the internet right now...")
  else:
      print(fun_fact_response)

4. Exact endpoint

This endpoint allows you to specify an exact endpoint to fetch.

.. code-block:: python

  from fetchable import FetchableClient

  client = FetchableClient()

  endpoint_response = client.fetch_endpoint("/v0.1/mount_everest/height")

  if(endpoint_response['status_code']==200):
      print("The height of mount_everest is {} {}s.".format(endpoint_response['value'], endpoint_response['unit']))
  elif(endpoint_response['status_code']==1001):
      print("Can't connect to the internet right now...")
  else:
      print(endpoint_response)


Errors
======

1. Exceptions

The FetchableClient object will throw exceptions when:

* The path to the authentication file is not set through an environment variable or constructor argument.
* The path to the authentication file is not a valid file path.
* The authentication file is not formatted properly.
* Functions taking string parameters are passed arguments which are not strings.

2. Error codes

There are two types of error codes you can receive back from the client object. Those in the 1xxx range are errors thrown by the client object itself and other error codes in the 2xx, 3xx, 4xx and 5xx ranges are the standard http error codes received from the server. For more information on the server error codes visit the `documentation <https://fetchable.ai/docs/api/general>`_.

+-------+------------------+--------------------------------------------------------------+
| Code  | Description      | Reason                                                       |
+-------+------------------+--------------------------------------------------------------+
| 1001  | Connection error |  The client cannot make a connection to the API server.      |
+-------+------------------+--------------------------------------------------------------+
| 1002  | Timeout error    |  The request timed out.                                      |
+-------+------------------+--------------------------------------------------------------+
| 1003  | Proxy error      |  A proxy error occurred.                                     |
+-------+------------------+--------------------------------------------------------------+
| 1004  | Unknown error    |  An unknown error occurred.                                  |
+-------+------------------+--------------------------------------------------------------+

Contributing
============

Contributions are welcome and encouraged! See the `Contributing Guide <CONTRIBUTING.rst>`_ for information on how to contribute.


License
=======
Licensed under Apache Version 2.0.

See the `LICENSE <LICENSE>`_ file for more information.


CHANGELOG.rst
==========
Change Log
==========

This document records all changes to the Fetchable Python client-side SDK and adheres to the `Semantic Versioning <https://semver.org/>`_ philosophy.

Notes
=====
Both the API and the client-side SDKS are versioned. The major and minor versions of client-side SDKs follow changes on the API side. e.g.

* 0.1.x for client-side SDKs corresponds to the v0.1 (Alpha) version of the API.
* 0.2.x for client-side SDKs corresponds to the v0.2 (Beta) version of the API.

The patch versions of the client-side SDKs (e.g. x.x.1, x.x.2, x.x.3, etc.) are independent of backend API versions and are free to update as they wish.

0.1 (Alpha version)
====================


0.1.0 (Released Aug 25, 2019)
-----------------------------
Initial public release with:


* client object with: ability to set user-agent; constructs authentication headers; and exposes wrapper functions for all endpoints in v0.1 of the backend API,

* example code,

* configuration class with API version selection,

* license, readme, contributing file, setup.py and requirements file.


0.1.1 (Released Sept 5, 2019)
-----------------------------
* Added compatibility between Python 2.7 and 3.6.


0.1.2 (Released Sept 6, 2019)
-----------------------------
* Cheekily changed public-facing interface since still in Alpha and no users will be affected - removed 'random' from fun fact, joke and quote functions.
* Added a change log and updated `Contributing Guide <CONTRIBUTING.rst>`_ guide.


0.1.3 (Released Sept 6, 2019)
-----------------------------
* Bugfixes in fetchable client object.
* Fixed errata in readme and changelog.


0.1.4 (Released Sept 7, 2019)
-----------------------------
* Updated Github issue templates.
* Added sanitisation of function parameters through private helper function, e.g. fetch_entity_atrribute("Empire State building", "Height") -> fetch_entity_atrribute("empire_state_building", "height").

0.1.5 (Released Sept 9, 2019)
-----------------------------
* Changed public API by fixing misspelling in function, not making this a major change because we're in Alpha stage with no customers.
* Changed project urls in setup.py.


fetchable/fetchable.py
import six

import os
import json
import requests
import datetime

import hashlib
import hmac
import base64

from .configuration import configuration


class FetchableClient(object):
    """
    The Fetchable API Client object.

    Use this object to interact with the Fetchable API.
    """


    def __init__(self, api_version=None, user_agent=None, auth_file=None):
        """
        Constructs Fetchable API Client object.

        The client automatically reads in the auth file on initialisation, it will throw an
        exception if cannot do this.

        :param api_version: The version of the Fetchable API to use.
        :type api_version: enum

        :param user_agent: The user-agent header to be sent to Fetchable.
        :type user_agent: string

        :param auth_file: The absolute path to the authentication file.
        :type auth_file: string
        """

        from .__version__ import __version__
        self.api_version = api_version or configuration.api_version.latest
        self.version = __version__
        self.user_agent = user_agent or 'fetchable-python-client/{}'.format(self.version)
        self.host = 'https://api.fetchable.ai'



        auth_file_path = auth_file or os.environ.get('FETCHABLE_AUTH_FILE')
        if not auth_file_path:
            raise Exception('path authentication file not set')

        if not os.path.isfile(auth_file_path):
            raise Exception('{} is not a valid authentication file path'.format(auth_file_path))

        credentials_file = open(auth_file_path)
        try:
            self.auth_credentials = json.load(credentials_file)
            credentials_file.close()
        except:
            raise Exception('authentication file is not formatted properly')
        if 'api_key' not in self.auth_credentials:
            raise Exception('authentication file is not formatted properly')
        if 'api_secret' not in self.auth_credentials:
            raise Exception('authentication file is not formatted properly')
        if type(self.auth_credentials['api_key']) != six.text_type:
            raise Exception('authentication file is not formatted properly')
        if type(self.auth_credentials['api_secret']) != six.text_type:
            raise Exception('authentication file is not formatted properly')

    def status(self):
        """
        Gets the status of the Fetchable API
        """
        return self.make_request(None)


    def fetch_entity_attribute(self, entity, attribute):
        """
        Fetches an entity-attribute value from the index

        :param entity: The entity to fetch.
        :type entity: str

        :param attribute: The attribute of the entity to fetch.
        :type attribute: str
        """
        if not entity or not attribute:
            raise Exception('parameters cannot be null')

        if type(entity) != str or type(attribute) != str:
            raise Exception('parameters are not strings')


        entity = self.sanitise_parameter(entity)
        attribute = self.sanitise_parameter(attribute)

        resource = self.api_version+"/"+entity+"/"+attribute
        return self.make_request(resource)


    def fetch_word_definition(self, word):
        """
        Fetches the definition of a word

        :param word: The word to get the definition of.
        :type word: str
        """

        if not word:
            raise Exception('parameters can not be null')

        if type(word) != str:
            raise Exception('parameters are not strings')

        word = self.sanitise_parameter(word)

        resource = self.api_version+"/"+word+"/definition"
        return self.make_request(resource)


    def fetch_joke(self):
        """
        Fetches a joke
        """
        resource = self.api_version+"/random/joke"
        return self.make_request(resource)


    def fetch_quote(self):
        """
        Fetches a quote
        """
        resource = self.api_version+"/random/quote"
        return self.make_request(resource)


    def fetch_fun_fact(self):
        """
        Fetches a fun fact
        """
        resource = self.api_version+"/random/fun_fact"
        return self.make_request(resource)


    def fetch_endpoint(self, endpoint):
        """
        Makes a request against a specific endpoint.

        Note: it is up to the caller to construct the endpoint here. This will override the version set in the constructor.
        Note: no parameter sanitisation is performed.
        Note: must start with a /
        Examples:   /ameliorate/definition
                    /v0.1/amazon_river_length

        :param endpoint: The endpoint to make a request against.
        :type endpoint: str
        """
        if not endpoint:
            raise Exception('parameters can not be null')

        if type(endpoint) != str:
            raise Exception('parameters are not strings')

        return self.make_request(endpoint)




    """
    Private helper functions
    """


    def make_request(self, resource):
        """
        Makes request to Fetchable API and returns the response body.
        """
        try:
            if resource:
                url = self.host+resource
                response = requests.get(url, headers=self.construct_headers(resource))
            else: #get status endpoint
                response = requests.get(self.host)
            body = json.loads(response.content)
            body['status_code'] = response.status_code
            return body

        except requests.exceptions.ConnectionError:
            body={}
            body['status_code'] = 1001
            body['reason'] = "Connection error occured"
            return body

        except requests.exceptions.Timeout:
            body={}
            body['status_code'] = 1002
            body['reason'] = "Timeout error occured"
            return body

        except requests.exceptions.ProxyError:
            body={}
            body['status_code'] = 1003
            body['reason'] = "Proxy error occured"
            return body

        except:
            body={}
            body['status_code'] = 1004
            body['reason'] = "Unknown error occured"
            return body


    def construct_headers(self, resource):
        """
        Constructs headers for the request
        """
        date = datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3]+"Z"
        signature_plain = ('GET'+resource+date).encode();

        hashed = hmac.new(self.auth_credentials['api_secret'].encode('ascii','ignore'),
            signature_plain,
            hashlib.sha1)
        signature1 = hashed.hexdigest().encode()
        signature2 = base64.urlsafe_b64encode(signature1)
        header = {
            "authorization": "Fetchable "+self.auth_credentials['api_key']+":"+signature2.decode(),
            "date": date,
            "user-agent": self.user_agent
        }
        return header


    def sanitise_parameter(self, param):
        """
        Sanitises parameters before sending them to the API, e.g.
        if the arguments passed to the entity-attribute function are:
        fetch_entity_atrribute("Empire State building", "Height")
        it will transform to "empire_state_building" and "height" before sending.
        """

        return param.replace(" ", "_").lower()


examples/main.py
from fetchable import FetchableClient
from fetchable import configuration

try:
    client = FetchableClient(api_version=configuration.api_version.latest)
except:
    print("problem initialising Fetchable client object")
    exit()

"""
get the status of Fetchable API
"""
status_response = client.status()

if(status_response['status_code']==200):
    print("The Fetchable API is up - all systems are go")
elif(status_response['status_code']==1001):
    print("Can't connect to the internet right now...")
else:
    print(status_response)



"""
fetch an entity - attribute value
"""
entity_response = client.fetch_entity_attribute("Mount Everest", "Elevation")

if(entity_response['status_code']==200):
    print("The height of mount_everest is {} {}.".format(entity_response['value'], entity_response['unit']))
elif(entity_response['status_code']==1001):
    print("Can't connect to the internet right now...")
else:
    print(entity_response)



"""
fetch a word definition
"""
definition_response = client.fetch_word_definition("ameliorate")

if(definition_response['status_code']==200):
    print("The definition of {}: is {}".format("ameliorate", definition_response['meanings'][0]))
elif(definition_response['status_code']==1001):
    print("Can't connect to the internet right now...")
else:
    print(definition_response)



"""
fetch a joke
"""
joke_response = client.fetch_joke()

if(joke_response['status_code']==200):
    print("{} - {}".format(joke_response['setup'], joke_response['punchline']))
elif(joke_response['status_code']==1001):
    print("Can't connect to the internet right now...")
else:
    print(joke_response)


"""
fetch a quote
"""
quote_response = client.fetch_quote()

if(quote_response['status_code']==200):
    print("{} by {}".format(quote_response['quote'], quote_response['author']))
elif(quote_response['status_code']==1001):
    print("Can't connect to the internet right now...")
else:
    print(quote_response)

"""
fetch a fun fact
"""
fun_fact_response = client.fetch_fun_fact()

if(fun_fact_response['status_code']==200):
    print(fun_fact_response['fun_fact'])
elif(fun_fact_response['status_code']==1001):
    print("Can't connect to the internet right now...")
else:
    print(fun_fact_response)

"""
make a request by defining the endpoint explicity
"""
endpoint_response = client.fetch_endpoint("/v0.1/mount_everest/height")

if(endpoint_response['status_code']==200):
    print("The height of mount_everest is {} {}s.".format(endpoint_response['value'], endpoint_response['unit']))
elif(endpoint_response['status_code']==1001):
    print("Can't connect to the internet right now...")
else:
    print(endpoint_response)
